voter/src/main/java/org/terracotta/voter/ActiveVoter.java

```diff
@@ -353,14 +401,91 @@ public void stop() {
     this.voter.interrupt();
   }
 
+  private void pollForTopologyFromActive(ExecutorService executorService, Optional<Properties> connectionProps,
+                                         AtomicReference<ClientVoterManager> voteOwner, List<Future<?>> futures) {
+    futures.add(executorService.submit(() -> {
+      heartbeatThreads.put("FetchingTopology", Thread.currentThread());
+      try {
+        activeClientVoterManager.connect(connectionProps);
+        while (activeClientVoterManager.isConnected()) {
+          Set<String> newTopology = activeClientVoterManager.getTopology();
+          LOGGER.info("Topology is {}.", newTopology);
+          if (!existingTopology.equals(newTopology)) {
+            LOGGER.info("New Topology detected {}.", newTopology);
+            // Start heartbeating with new servers
+            List<String> addedServers = getAddedServers(newTopology);
+            addedServers.forEach(server -> {
+              existingTopology.add(server);
+              ClientVoterManager clientVoterManager = new ClientVoterManagerImpl(server);
+              voterManagers.add(clientVoterManager);
+              synchronized (registrationLatch) {
+                registrationLatch.add(server);
+              }
+              futures.add(executorService.submit(heartbeat(clientVoterManager, connectionProps, voteOwner)));
+            });
+
+            // Do removal of old servers from topology
+            List<String> removedServers = getRemovedServers(newTopology);
+            removedServers.forEach(server -> {
+              existingTopology.remove(server);
+              heartbeatThreads.get(server).interrupt();
+              heartbeatThreads.remove(server);
+            });
+          }
+          sleepForTopologyFetchInterval();
+        }
+      } catch (TimeoutException | RuntimeException e) {
+        activeClientVoterManager.close();
+        sleepFor10();
+      }
+    }));
+  }
+
+  private List<String> getRemovedServers(Set<String> newTopology) {
+    List<String> res = new ArrayList<>();
+    existingTopology.forEach(server -> {
+      if (!newTopology.contains(server)) {
+        res.add(server);
+      }
+    });
+    return res;
+  }
+
+  private List<String> getAddedServers(Set<String> newTopology) {
+    List<String> res = new ArrayList<>();
+    newTopology.forEach(server -> {
+      if (!existingTopology.contains(server)) {
+        res.add(server);
+      }
+    });
+    return res;
+  }
+
+  private void sleepForTopologyFetchInterval() {
+    try {
+      Thread.sleep(topologyFetchInterval);
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
   @Override
   public void close() {
     stop();
   }
 
   @Override
   public String toString() {
-    return "ActiveVoter{" + Arrays.toString(servers) + "}";
+    return "ActiveVoter{" + existingTopology + "}";
+  }
+
+  // For testing purposes
+  public Set<String> getExistingTopology() {
```

Comment:
package-private if this is just for testing.

Reply:
It will not be accessible then for tests in galvan module.